### PR TITLE
#16038 deprecate symbol ==> Symbol

### PR DESCRIPTION
**Pull request by jeffreysarnoff-dev**
deprecate symbol ==> Symbol (first commit)
minor fixups relating to symbol(string("..")) (second commit)

### CODE REVIEW

**Comment by jeffreysarnoff-dev** [comment]
replaces pull request #15995, wait for the second commit here.

**Comment by jeffreysarnoff-dev** [comment]
Second commit complete, more likely than not: it is good to go.

**Comment by nalimilan** [comment]
Thanks, looks good (though you could have changed all occurrences of `string(op)[2:end]`).

**Comment by stevengj** [comment]
(In general, it is better to do a forced push rather than opening a new PR, so that all the discussion goes into a single PR.)

**Comment by Jeffrey-Sarnoff** [comment]
@stevengj with practice [till then, the link exists]
I am content should this work.

File: base/expr.jl [hunk 1]
```diff
@@ -2,12 +2,12 @@
...
-symbol(s::ASCIIString) = symbol(s.data)
-symbol(s::UTF8String) = symbol(s.data)
-symbol(a::Array{UInt8,1}) =
+Symbol(s::Symbol) = s
+Symbol(s::ASCIIString) = Symbol(s.data)
+Symbol(s::UTF8String) = Symbol(s.data)
+Symbol(a::Array{UInt8,1}) =
     ccall(:jl_symbol_n, Ref{Symbol}, (Ptr{UInt8}, Int32), a, length(a))
-symbol(x...) = symbol(string(x...))
+Symbol(x...) = Symbol(string(x...))
```
Comment (nalimilan) [inline base/expr.jl]: The test failures come from this: a test in replutil.jl:227 expects `Symbol()` to throw a `MethodError` (https://github.com/JuliaLang/julia/pull/14884), and with this change it returns `Symbol("")`. I'd say we should just remove the test, but that part of the code is quite obscure.

@vtjnash @tkelman What's the best solution here?

Comment (vtjnash) [inline base/expr.jl]: yes, this code replaces that test (which was testing that adding methods to `Symbol` worked correctly)

Comment (Jeffrey-Sarnoff) [inline base/expr.jl]: So, there is nothing for me to do? (he asks hopefully)

**Comment by jeffreysarnoff-dev** [comment]
There was one fixup missed in threadcall.jl. That is fixed, it may be a separate commit.  The only case of [2:end] that I did not change should not be changed afaik.

File: base/dates/periods.jl [hunk 2]
```diff
@@ -229,7 +229,7 @@ GeneralPeriod = Union{Period,CompoundPeriod}
 (+){P<:GeneralPeriod}(x::StridedArray{P}) = x
 
 for op in (:.+, :.-)
-    op_ = symbol(string(op)[2:end])
+    op_ = Symbol(string(op)[2])
```
Comment (tkelman) [inline base/dates/periods.jl]: I don't think this is right

Comment (jeffreysarnoff-dev) [inline base/dates/periods.jl]: You are correct, the edit is undone.

File: base/expr.jl (hunk 1 above)
Comment (nalimilan) [inline base/expr.jl]: @vtjnash can confirm whether that's what he meant, but at the minimum you should change the test I referred to. For example, replace `Symbol` with `sum()` (be sure to change the error message check too).

Then, please squash the changes into 2 commits. You can do that using `git rebase -i master`, reordering the commits as needed, and then writing `squash` for those you want to merge with the preceding commit. Then, do `git push jeffreysarnoff-dev +jas/deprecate_symbol` to update the PR.

**Comment by jeffreysarnoff-dev** [comment]
Is there anything I can to do to make this ready?  If the stuff in the second commit reveals other issues, should I adjust this so that only the first commit is present, leaving the second commit to some other PR?

**Comment by stevengj** [comment]
You'll want to merge the last three commits (via `git rebase -i HEAD~4` followed by a forced push) so that you only have two commits: one for the deprecation and one for the cleanup.

You'll also need to rebase against master to fix the conflicts.

And finally, the tests are failing with `err not defined`, so you'll need to fix that.

**Comment by jeffreysarnoff-dev** [comment]
@stevengj I don't know how to do that.  I know how to change this so that it only does symbol(x...) ==> Symbol. And I know how to say "I don't know how to do that."

**Comment by stevengj** [comment]
See e.g. the instructions here: https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request

**Comment by jeffreysarnoff-dev** [comment]
trying anyway

**Comment by nalimilan** [comment]
See also my detailed instructions at https://github.com/JuliaLang/julia/pull/16038#discussion_r61049969.

But note that the fix for the failing tests will have to go into the *first* commit, so that it doesn't break the build.

**Comment by jeffreysarnoff-dev** [comment]
I may have "You'll want to merge the last three commits (via git rebase -i HEAD~4 followed by a forced push) so that you only have two commits: one for the deprecation and one for the cleanup. You'll also need to rebase against master to fix the conflicts."

In the event that that went as intended, and the fix is to eliminate the test (or, just avoid the conflict), How do I put that into the first commit?

**Comment by nalimilan** [comment]
Commit the change, do `git rebase -i master`, move the commit with the tests fix to the line below your first commit, mark it as `squash`. Then you need to force-push again.

Note your last push removed your commits and added unrelated ones...

**Comment by jeffreysarnoff-dev** [comment]
Thank you for the help -- whatever I did did not make the repository as it should be,
even though I did the rebasing and forced push. I will revisit this tonight.  There are two different things in this PR now, and the original intent works;  the second part, replacing Symbol(string(a),...) with Symbol(a,...) is where the conflicts happen and where my lack of familiarity with (or trust in) git complicates this.  I will reread the guidance and retry for both, if it not work -- I am going to provide jas/deprecate_symbol as current with julia + 1 commit.

**Comment by nalimilan** [comment]
If you can make this PR pass the tests with only the first part, that's OK. Then you'll be able to open a new PR for the improvements.

**Comment by jeffreysarnoff-dev** [comment]
updated to a single commit just replacing symbol with Symbol in .jl, .md, .rst
and keeping _symbol as _symbol
https://github.com/jeffreysarnoff-dev/julia/tree/jas/deprecate_symbol

**Comment by jeffreysarnoff-dev** [comment]
replaced by clean version PR 16130

File: base/expr.jl (hunk 1 above)
Comment (KristofferC) [reply base/expr.jl]: This was never done?

**Pull request closed without merging**